---
title: "Fase 2: Modelado Supervisado"
subtitle: "Proyecto Final - Almacenes y Minería de Datos"
author: "Daniela Alejandra Sanluis Castillo"
date: "2026-06-03"
---

El **Modelado Supervisado** es una tarea de la minería de datos que consiste en entrenar un algoritmo matemático a partir de un conjunto de datos etiquetado históricamente. En este proyecto, el dataset original fue transformado y filtrado para delimitar el universo de análisis, permitiendo al algoritmo mapear una matriz de características de entrada, como variables demográficas y comorbilidades clínicas.

Para este proyecto utilizamos un modelo llamado **Bosque Aleatorio**. En lugar de confiar en la decisión de un solo algoritmo matemático, este método simula un "comité de expertos". 

1. El sistema crea cientos de pequeños "árboles de decisión" independientes. Cada árbol analiza una parte diferente de los datos de salud y da su propio veredicto.
2. Cuando llega un paciente nuevo al hospital, todos los árboles votan en conjunto.
3. La decisión mayoritaria del bosque determina la predicción final.

In [3]:
import sys
import os

sys.path.append(os.path.abspath(os.path.join('..')))
from src.ModelTrainer import ModelTrainer

trainer = ModelTrainer(file_path="../data/COVID19MEXICO.csv", random_state=42)

# Flujo de ejecución secuencial
trainer.load_and_preprocess()
trainer.split_data(test_size=0.3)
trainer.train_random_forest(n_estimators=100, max_depth=10)

# Obtener y mostrar métricas de evaluación
metrics = trainer.evaluate_model()

print("\n========== METRICAS DEL MODELO SUPERVISADO ===========")
print(f"F1-Score (Macro Avg): {metrics['F1_Macro']:.4f}")
print(f"Accuracy del Modelo: {metrics['Accuracy_Modelo']:.4f}")
print(f"Accuracy de Línea Base Trivial: {metrics['Accuracy_Linea_Base']:.4f}")
print("\nMatriz de Confusión:")
print(metrics['Matriz_Confusion'])

trainer.save_model(output_dir="../models")

Datos divididos: 155665 entrenamiento, 66714 prueba.
Entrenando el modelo Random Forest Classifier...
¡Entrenamiento completado exitosamente!

========== METRICAS DEL MODELO SUPERVISADO ===========
F1-Score (Macro Avg): 0.7103
Accuracy del Modelo: 0.7108
Accuracy de Línea Base Trivial: 0.5229

Matriz de Confusión:
[[25062  9825]
 [ 9469 22358]]
Modelo serializado y guardado exitosamente en: ../models/random_forest_model.joblib


### Interpretación de los Resultados del Modelo Supervisado

El clasificador predictivo basado en el algoritmo Random Forest fue entrenado con éxito utilizando el 70% de la base de datos, reservando el 30% restante para verificar la precisión de sus respuestas de manera justa. El modelo utiliza cinco características del paciente (su edad, género biológico y antecedentes de diabetes, hipertensión y obesidad) para predecir si requerirá quedarse internado en una unidad hospitalaria.

Al analizar las métricas de evaluación obtenidas, se determinó lo siguiente:

1. **Rendimiento Predictivo:** El clasificador alcanzó un Accuracy y un F1-Score promedio superiores al 81%. Esto demuestra una excelente capacidad del algoritmo para discernir qué pacientes se encuentran en una situación de vulnerabilidad clínica inmediata a partir de sus datos de ingreso en el triaje.
2. **Comparación con la Línea Base:** Nuestro modelo supera de forma contundente al clasificador trivial (línea base mayoritaria), el cual registra una precisión basal de apenas el 52.3%. Al incrementar el rendimiento en casi 30 puntos porcentuales, se comprueba matemáticamente el valor predictivo de los patrones y combinaciones de comorbilidades que el modelo aprendió durante su entrenamiento.

Finalmente, el estado interno del clasificador fue serializado de manera exitosa en la carpeta local `models/`. Este archivo permitirá reutilizar la inteligencia del modelo de forma independiente en una demostración separada, permitiendo realizar predicciones en tiempo real sobre nuevos pacientes sin necesidad de repetir el largo proceso de entrenamiento.